readthedocs/proxito/views/mixins.py

```diff
@@ -154,3 +154,38 @@ def get_redirect_response(self, request, redirect_path, http_status):
             return HttpResponsePermanentRedirect(new_path)
 
         return HttpResponseRedirect(new_path)
+
+
+class ServeDocsPermissionsMixin:
+
+    """
+    Mixin to handle documentation permissions.
+
+    To use this mixin, the method that serve the docs should call
+    ``allowed_user`` first to check if the user requesting to read the page is
+    allowed. If allowed, the flow should continue.
+
+    On the other hand, if the user is not allowed, the serve docs method MUST
+    break the flow and return a call to ``get_unauthed_response``.
+    """
+
+    def allowed_user(self, request, project, version_slug):
+        """
+        Return whether the user is allowed to read this documentation's version.
+
+        :returns: ``True`` if allowed. ``False`` otherwise.
+        :rtype: bool
+        """
+        return False
+
+    def get_unauthed_response(self, request, project):
+        """
+        Return response for an unauthed hit at a documentation's version.
+
+        When a user does not have access to read documentation, we need to
+        return a response showing the proper message. This method produces that
+        reponse.
+
+        :rtype: django.http.HttpResponse
+        """
+        return self._serve_401(request, project)
```

Comment:
I don't quite understand what this is meant to be doing. Are we using these additional methods, or are they just meant to be overwritten?

Reply:
We are overriding these methods from corporate. They are meant to be overwritten.

This PR just move them outside the main class and put them into a mixin with better docstrings. Which is easier to understand where they come from. That's all. 

I could move the docstrings to the Base class, otherwise.

Reply:
@humitos I think I ended up doing most of this PR in https://github.com/readthedocs/readthedocs.org/pull/6572 -- we should get this merged and combined with that one 👍

Reply:
@ericholscher see https://github.com/readthedocs/readthedocs.org/pull/6572/files#r371770894

I'm fine with either. I think defaulting to `False` and force re-define it, it's safer, though.

Reply:
Yea, it should default False 👍